### ISSUE TITLE
Can I used it in-place of FastSpeech's Length Regulator 

### ISSUE THREAD

**Issue by rishikksh20** [Joshua-1995/LearnableUpsamplingLayer-Pytorch#1]
@Joshua-1995 
I haven't read the Natural Speech paper, though I am curious to know if this Upsample can replace vanilla repeatability-based Length Regulator which is usually used in FastSpeech like Non-Autoregressive TTS system.

**Comment by Joshua-1995** [comment]
Hi, @rishikksh20.

I'm glad to have a discussion about this.

As I know, the technique was introduced in Parallel Tactron 2[1], which is a recently proposed non-AR TTS.
The main advantage of using the differentiable model is enabling the duration predictor to be updated by gradient of the main objective function (spectrogram reconstruction loss).

For the FastSpeech-like non-AR TTS systems, the phoneme duration needs to be rounded before expanding the phoneme representation to frame-level representation. Since the rounding operation is not differentiable, it prevents the joint optimization of the duration predictor and the decoder so as to minimize the reconstruction loss [1].

Furthermore, FastSpeech-like non-AR TTS needs to be trained in a teacher-forcing with ground-truth duration to match the length of target data. This also prevents the gradient of reconstruction loss from going through the duration predictor.

To tackle the challenge, the paper proposed two techniques to enable the duration predictor to be updated by the gradient of spectrogram reconstruction loss. The first one is the differentiable length regulator, which makes the length regulating process differentiable. The second one is soft-DTW, which enables TTS systems to be trained without teacher forcing of target duration.

Therefore, in order to take full advantage of using the differentiable duration model, which is enabling the duration predictor to be updated by the main objective, we need to use a soft-DTW version of reconstruction loss. 

However, in my opinion, the differentiable duration model without soft-DTW loss would be also helpful.
The differentiable duration model uses a attention-like modules thus outputs soft-alignment. This might reflect the soft transition of phoneme sound, which is more reasonable than using the hard alignment. 
For example, let us consider two adjacent phonemes and their sound. The first comming sound doesn't disapper immediately but gradually.

[1] Isaac Elias, Heiga Zen, Jonathan Shen, Yu Zhang, Jia Ye, RJ Ryan, and Yonghui Wu. Parallel
tacotron 2: A non-autoregressive neural tts model with differentiable duration modeling. arXiv
preprint arXiv:2103.14574, 2021.
